### ISSUE TITLE
Is it meaningful to let neutral network model analyze less vertices to speedup computing?

### ISSUE THREAD

**Issue by HackYardo** [lightvector/KataGo#845]
There are many vertices and series in analyze outputs, but in fact the top three vertices and their series are most be used by real human Go players. How to limit analyze to consider less vertices? 

#103 
I find it that the `maxmoves` command flag is only "max vertices to display" but not "max vertices to consider".
```
./katago gtp
analyze 1000 maxmoves 3
analyze 1000 
quit
```
It seems here no visits amounts differences between mamoves or not.
And there is a `analysisWideRootNoise` but it is made to consider more vertices not less.
Third, the `allow PLAYER VERTEX,VERTEX,... UNTILDEPTH`, but it is for a specified serie only.

**Comment by lightvector** [comment]
That's an interesting suggestion. I don't see how this could be a good idea. The network and/or the MCTS search don't "know" which 3 moves will be the best 3 at the end of the search. So if you require the bot to prematurely pick 3 moves and only analyze those 3, there will be frequently cases where the 3 moves chosen won't contain the best move, and then you are in trouble - no matter how much analysis you do after that it will never be able to correct the error.

**Comment by HackYardo** [comment]
I heard it that the **raw network (1 visit per turn)** is strong enough to beat 6D Go player. So I want to call KataGo in this way:
1. use a bit visits to get the top 3 vertices
2. use the most visits to analyze the top 3 only

Perhaps parallel computing is needed for step 2, could I do it in gtp mod, or I have to do it in json protocol (parallel analysis engine)?

**Comment by OmnipotentEntity** [comment]
You could do this, but it likely would harm the strength of the bot at high visits.

**Comment by awsjgy** [comment]
> I heard it that the **raw network (1 visit per turn)** is strong enough to beat 6D Go player. So I want to call KataGo in this way:
> 
> 1. use a bit visits to get the top 3 vertices
> 2. use the most visits to analyze the top 3 only (send the 3 positions to engine and then analyze)
> 
> Perhaps parallel computing is needed for step 2, could I do it in gtp mod, or I have to do it in json protocol (parallel analysis engine)?

I've thought about this suggestion of yours, and there's nothing wrong with it for those machines with higher configurations. But for most home computers, which are not that well equipped, it is true that reducing the number of selected points can serve the theoretical purpose of increasing the speed and accuracy of the calculation. But I wonder if katago supports this algorithm. Need to further optimize
你的这个建议我也想过，对于那些配置比较高的机器来说无可厚非。但是对于大部分家用电脑，配置没那么好，的确减少选点可以起到提高计算速度精度的理论意义。但是不知道katago支持这种算法吗。需要进一步优化

**Comment by HackYardo** [comment]
I tried, it works, thank you! And algorithm performance improvement is beyond my border. I seek efficiency so many years just because I'm on old weak hardware ;D
```
root@localhost:~/KataGo# ./katago gtp
analyze 100 maxmoves 3
=
info move Q16 visits 13 utility 0.0457608 winrate 0.530481 scoreMean 2.46073 scoreStdev 26.4767 scoreLead 2.46073 scoreSelfplay 1.24387 prior 0.0166064 lcb 0.505290 utilityLcb -0.0247750 weight 13.0000 order 0 pv Q16 C17 Q3 C3 Q5 info move Q4 visits 13 utility 0.0457608 winrate 0.530481 scoreMean 2.46073 scoreStdev 26.4767 scoreLead 2.46073 scoreSelfplay 1.24387 prior 0.0166064 lcb 0.505290 utilityLcb -0.0247750 weight 13.0000 isSymmetryOf Q16 order 1 pv Q4 C3 Q17 C17 Q15 info move D16 visits 13 utility 0.0457608 winrate 0.530481 scoreMean 2.46073 scoreStdev 26.4767 scoreLead 2.46073 scoreSelfplay 1.24387 prior 0.0166064 lcb 0.505290 utilityLcb -0.0247750 weight 13.0000 isSymmetryOf Q16 order 2 pv D16 R17 D3 R3 D5


analyze 1000 allow b q16,q4,d16 1
=
quitinfo move Q16 visits 1419 utility 0.0218856 winrate 0.513202 scoreMean 1.31044 scoreStdev 26.7675 scoreLead 1.31044 scoreSelfplay 0.460957 prior 0.0166064 lcb 0.511140 utilityLcb 0.0161130 weight 1418.36 order 0 pv Q16 C4 R3 C16 E3 R17 Q17 R16 R15 S15 info move Q4 visits 1419 utility 0.0218856 winrate 0.513202 scoreMean 1.31044 scoreStdev 26.7675 scoreLead 1.31044 scoreSelfplay 0.460957 prior 0.0166064 lcb 0.511140 utilityLcb 0.0161130 weight 1418.36 isSymmetryOf Q16 order 1 pv Q4 C16 R17 C4 E17 R3 Q3 R4 R5 S5 info move D16 visits 1419 utility 0.0218856 winrate 0.513202 scoreMean 1.31044 scoreStdev 26.7675 scoreLead 1.31044 scoreSelfplay 0.460957 prior 0.0166064 lcb 0.511140 utilityLcb 0.0161130 weight 1418.36 isSymmetryOf Q16 order 2 pv D16 R4 C3 R16 P3 C17 D17 C16 C15 B15


=

```

**Comment by lightvector** [comment]
@HackYardo - I would caution you that unless you have rigorously tested it and proven that it is stronger, your approach has a good chance to be *inefficient*, even on weak hardware. And by "inefficient" i mean that even on weak hardware, the same amount of computation spent with such a method is likely to produce a weaker final result and analysis than just running without such a method and searching more moves.

**Comment by HackYardo** [comment]
Yeah, I understand, I have learned a lot from the issues I opened. Thank you again for your caution and your repo. The focused computing sources may not produce better outputs but even worse, so the focus is inefficient. I won't use `allow` before I test it massively, I wil use `maxmoves` instead.

在 2023-10-17 08:22:10，lightvector ***@***.***> 写道：


@HackYardo - I would caution you that unless you have rigorously tested it and proven that it is stronger, your approach has a good chance to be inefficient, even on weak hardware. And by "inefficient" i mean that even on weak hardware, the same amount of computation spent with such a method is likely to produce a weaker final result and analysis than just running without such a method and searching more moves.
